source/docs/flow_control.md

```diff
@@ -0,0 +1,182 @@
+buffer contains more data than the configured limit, the high watermark callback will fire, kicking
+off a chain of events which eventually informs the data source to stop sending data.  This back-off
+may be immediate (stop reading from a socket) or gradual (stop HTTP/2 window updates) so all
+buffer limits in Envoy are considered soft limits.  When the buffer eventually drains (generally to
+half of of the high watermark to avoid thrashing back and forth) the low watermark callback will
+fire, informing the sender it can resume sending data.
+
+### TCP implementation details
+
+Flow control for TCP and TCP-with-TLS-termination are handled by coordination
+between the `Network::ConnectionImpl` write buffer, and the `Network::TcpProxy`
+filter.
+
+The downstream flow control goes as follows.
+
+ * The downstream `Network::ConnectionImpl::write_buffer_` buffers too much
+   data.  It calls
+   `Network::ConnectionCallbacks::onAboveWriteBufferHighWatermark()`.
+ * The `Network::TcpProxy::DownstreamCallbacks` receives
+   `onAboveWriteBufferHighWatermark()` and calls `readDisable(true)` on the upstream
+   connection.
+ * When the downstream buffer is drained, it calls
+   `Network::ConnectionCallbacks::onBelowWriteBufferLowWatermark()`
+ * The `Network::TcpProxy::DownstreamCallbacks` receives
+   `onBelowWriteBufferLowWatermark()` and calls `readDisable(false)` on the upstream
+   connection.
+
+Flow control for the upstream path is much the same.
+
+ * The upstream `Network::ConnectionImpl::write_buffer_` buffers too much
+   data.  It calls
+   `Network::ConnectionCallbacks::onAboveWriteBufferHighWatermark()`.
+ * The Network::TcpProxy::UpstreamCallbacks receives
+   `onAboveWriteBufferHighWatermark()` and calls `readDisable(true)` on the downstream
+   connection.
+ * When the upstream buffer is drained, it calls
+   `Network::ConnectionCallbacks::onBelowWriteBufferLowWatermark()`
+ * The `Network::TcpProxy::UpstreamCallbacks` receives
+   `onBelowWriteBufferLowWatermark()` and calls `readDisable(false)` on the downstream
+   connection.
+
+### HTTP2 implementation details
+
+Because the various buffers in the HTTP/2 stack are fairly complicated, each path from a buffer
+going over the watermark limit to disabling data from the data source is documented separately.
+
+![HTTP2 data flow diagram](h2_buffers.png)
+
+For HTTP/2, when filters, streams, or connections back up, the end result is `readDisable(true)`
+being called on the source stream.  This results in the stream ceasing to consume window, and so
+not sending further flow control window updates to the peer.  This will result in the peer
+eventually stopping sending data when the available window is consumed (or nghttp2 closing the
+connection if the peer violates the flow control limit).  When `readDisable(false)` is called, any
+outstanding unconsumed data is immediately consumed, which results in resuming window updates to the
+peer and the resumption of data.
```

Comment:
Might be worth pointing out that we only do H2 flow control on stream level and not the connection level for the purposes of buffer bounding, which is implied here but a useful thing to call out explicitly.